vici/session.go

```diff
@@ -152,7 +164,29 @@ func newFuncSessionOption(f func(*sessionOpts)) *funcSessionOption {
 // path, /var/run/charon.vici, is used.
 func WithSocketPath(path string) SessionOption {
 	return newFuncSessionOption(func(so *sessionOpts) {
-		so.path = path
+		so.network = "unix"
+		so.addr = path
+	})
+}
+
+// WithAddr specifies the network and address of the socket that charon
+// is listening on. If this option is not specified, the default
+// path, /var/run/charon.vici, is used.
+//
+// As the protocol itself currently does not provide any security or
+// authentication properties, it is recommended to run it over a UNIX
+// socket with appropriate permissions.
+func WithAddr(network, addr string) SessionOption {
+	return newFuncSessionOption(func(so *sessionOpts) {
+		so.network = network
+		so.addr = addr
+	})
+}
+
+// WithDialer specifies the dial func to use when dialing the charon socket.
+func WithDialer(dialer func(ctx context.Context, network, addr string) (net.Conn, error)) SessionOption {
```

Comment:
Do you have a specific use case for this right now? As-is, I don't see why this option is helpful (particularly because `so.dialer` is only ever invoked with `context.Background()`). Are you constructing a `net.Dialer` with a timeout/deadline set?

Reply:
So I would like to apply a `context.Context` to the entire constructions of the `vici.Session`, including the part where we dial the socket. Since `NewSession` does blocking work to dial the conn, ideally it would take a `context.Context`. I didn't want to combine those API changes in this PR, so my plan was to use the custom dialer function and use my own `context.Context`. 

I copied this pattern from `http.Transport`, where you can specify a `DialContext` method, and grpc-go's [`WithContextDialer`](https://github.com/grpc/grpc-go/blob/506b7730668b5a13465224b0d8133f974a3f843d/dialoptions.go#L363).

Reply:
Let me know what you think - if you would like me to get rid of this method, then happy to accept the changes you suggest above.

Reply:
Okay, that's what I thought you were going for, but in those APIs it's always been a little unclear to me what `context.Context` actually gets passed into `DialContext`, and if `DialContext` is supposed to then build on that context. Anyways, is the idea that you will do something like:

```go
dialer := net.Dialer{Timeout: 3 * time.Second}
s, err := vici.NewSession(vici.WithDialer(dialer.DialContext))
// ...
```
and then due to https://golang.org/src/net/dial.go#L120, your context will have a timeout even though `vici.NewSession` calls `s.dialer(context.Background(), s.network, s.addr)`?

Reply:
If that's the case (or you define your own function that wraps `context.Background` to to something more useful), I think this makes sense to add. I just needed to see how this would be used.

Reply:
Yup the later is what I was planning to do for my use case. I.e.

```
func doSomeWork(ctx context.Context) {
  s, err := vici.NewSession(vici.WithDialer(func(_ context.Context, network, addr string) (net.Conn, error) {
    return net.DialContext(ctx, network, addr)
  }))
  ...
}
```

This is a workaround really.. But works for me!

Reply:
Alright, let's do this:

1. Please `s/WithDialer/WithDialContext` for consistency/clarity.
2. Remove the global `var defaultDialer` (and use my other suggestion)

And I'm good to merge :+1:

Reply:
sweet! done!